Meta/serenity.sh

```diff
@@ -0,0 +1,190 @@
+#!/bin/env bash
+set -e
+
+ARG0=$0
+function print_help()
+{
+  cat <<EOF
+Usage: $ARG0 COMMAND [TARGET] [ARGS...]
+  Supported COMMANDs:
+    build:     Compiles the target binaries, [ARGS...] are passed through to ninja
+    install:   Installs the target binary
+    image:     Creates a disk image with the installed binaries
+    run:       TARGET lagom:      Runs the unit tests on the build host
+               All other TARGETs: $ARG0 run [TARGET] [kernel_cmd_line]
+                   Runs the built image in QEMU, and optionally passed the
+    delete:    Removes the build environment for the TARGET
+    recreate:  Deletes and re-creates the build environment for TARGET
+    rebuild:   Deletes and re-creates the build environment, and compiles for TARGET
+    addr2line: $ARG0 addr2line TARGET BINARY_FILE ADDRESS
+                   Resolves the ADDRESS in BINARY_FILE to a file:line
+
+    rebuild-toolchain: Deletes and re-builds the TARGET's toolchain
+  Supported TARGETs: i686 (default), x86_64, lagom
+
+  Examples:
+    $ARG0 run i686 smp=on
+        Runs the image in QEMU passing "smp=on" to the kernel command line
+    $ARG0 run
+        Runs the image for the default TARGET i686 in QEMU
+    $ARG0 run lagom
+        Runs the unit tests on the build host
+    $ARG0 addr2line i686 Kernel/Kernel 0x12345678
+        Resolves the address 0x12345678 in the binary Kernel/Kernel
+EOF
+}
+
+function usage()
+{
+  >&2 print_help
+  exit 1
+}
+
+CMD=$1;
+[ ! -z "$CMD" ] || usage
+shift
+if [ $CMD = "help" ]; then
+  print_help
+  exit 0
+fi
+
+if [ ! -z "$1" ]; then
+  TARGET="$1"; shift
+else
+  TARGET="i686"
+fi
+CMD_ARGS=( "$@" )
+CMAKE_ARGS=()
+
+function get_top_dir()
+{
+  git rev-parse --show-toplevel
+}
+
+function is_valid_target()
+{
+  if [ $TARGET = "lagom" ]; then
+    echo 'target is lagom';
+    CMAKE_ARGS+=("-DBUILD_LAGOM=ON")
+    return 0
+  fi
+  [[ $TARGET =~ ^(i686|x86_64|lagom)$ ]] || return 1
+}
+
+function create_build_dir()
+{
+  mkdir -p "$BUILD_DIR"
+  ( cd "$BUILD_DIR" && cmake -GNinja "${CMAKE_ARGS[@]}" ../..)
+}
+
+function cmd_with_target()
+{
+  is_valid_target || ( >&2 echo "Unknown target: $TARGET"; usage )
+  export SERENITY_ARCH="$TARGET"
+  SERENITY_ROOT=$(get_top_dir)
+  export SERENITY_ROOT
+  BUILD_DIR="$SERENITY_ROOT/Build/$TARGET"
```

Comment:
I like the new directory layout, but I think the Toolchain-builder has clashing assumptions built into it. Currently, `BuildIt.sh` contains the line:
```
BUILD="$DIR/../Build"
```

Maybe it is sufficient to change this to:
```
BUILD="$DIR/../Build/$ARCH"
```

I'm not sure whether this is all it takes, so it might be a good idea to run the commands locally, which ensures that your changes really work. :)

Reply:
Hm interesting. I like your suggestion.